cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -0,0 +1,239 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gtest/gtest.h>
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/kernels/test_util.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/formatting.h"
+
+namespace arrow {
+
+using internal::StringFormatter;
+
+class ScalarTemporalTest : public ::testing::Test {};
+
+namespace compute {
+
+TEST(ScalarTemporalTest, TestSimpleTemporalComponentExtraction) {
+  const char* json =
+      R"(["1970-01-01T00:00:59","2000-02-29T23:23:23",
+          "3989-07-14T18:04:01","1900-01-01T01:59:20","2033-05-18T03:33:20"])";
+  auto time_points = ArrayFromJSON(timestamp(TimeUnit::SECOND), json);
+
+  auto year = ArrayFromJSON(int64(), "[1970, 2000, 3989, 1900, 2033]");
+  auto month = ArrayFromJSON(int64(), "[1, 2, 7, 1, 5]");
+  auto day = ArrayFromJSON(int64(), "[1, 29, 14, 1, 18]");
+  auto day_of_year = ArrayFromJSON(int64(), "[1, 60, 195, 1, 138]");
+  auto week = ArrayFromJSON(int64(), "[1, 9, 28, 1, 20]");
+  auto quarter = ArrayFromJSON(int64(), "[1, 1, 3, 1, 2]");
+  auto day_of_week = ArrayFromJSON(int64(), "[4, 2, 5, 1, 3]");
+  auto hour = ArrayFromJSON(int64(), "[0, 23, 18, 1, 3]");
+  auto minute = ArrayFromJSON(int64(), "[0, 23, 4, 59, 33]");
+  auto second = ArrayFromJSON(int64(), "[59, 23, 1, 20, 20]");
+  auto millisecond = ArrayFromJSON(int64(), "[0, 0, 0, 0, 0]");
+  auto microsecond = ArrayFromJSON(int64(), "[0, 0, 0, 0, 0]");
+  auto nanosecond = ArrayFromJSON(int64(), "[0, 0, 0, 0, 0]");
```

Comment:
Is it desirable for all these values to be `int64`? A number of hours, minutes or seconds should fit in a `int8`, for example.